c2rust-analyze/src/util.rs

```diff
@@ -307,19 +307,23 @@ pub fn lty_project<'tcx, L: Debug>(
     }
 }
 
-/// Determine if two types are safe to transmute to each other.
+/// Determine if `from_ty` can be safely transmuted to `to_ty`.
 ///
 /// Safe transmutability is difficult to check abstractly,
-/// so here it is limited to integer types of the same size
-/// (but potentially different signedness).
+/// so here it is limited to
+/// * integer types of the same size (but potentially different signedness)
+///     * e.x. `*u8 as *i8`
+/// * decaying arrays and slices to their element type
+///     * e.x. `*[u8; 0] as *u8`
+///     * e.x. `*[u8] as *u8`
 ///
 /// Thus, [`true`] means it is definitely transmutable,
 /// while [`false`] means it may not be transmutable.
-pub fn are_transmutable<'tcx>(a: Ty<'tcx>, b: Ty<'tcx>) -> bool {
-    let transmutable_ints = {
+pub fn is_transmutable_to<'tcx>(from_ty: Ty<'tcx>, to_ty: Ty<'tcx>) -> bool {
+    let transmutable_ints = || {
```

Comment:
This documentation isn't completely correct now that the function allows `from_ty` to decay: for example, `is_transmutable_to([u8; 4], u8)` returns `true`, but you can't transmute from `[u8; 4]` to `u8` in Rust. I think the docs and name should be updated to clarify that this only applies behind a pointer, and perhaps it makes the most sense to simply fold this into `is_transmutable_ptr_cast`, as that's the only caller as well as the necessary context to understand the predicate being applied.

Reply:
`do_unify` also calls `is_transmutable_to`, that's why they're separate.  I'll fix the docs, though, and clarify that it's for behind a pointer, or more similar to `transmute_copy` than `transmute` now.

Reply:
The docs have been fixed/updated and the definition formalized in b84b4129cfa2e85fd910932bed795434e01c66f3.